File: Bài toán mô phỏng.py
# code by Lê Văn Long - B22DCVT316

import heapq

INF = int(1e9)
maxn = 26  

n, m = 0, 0
s = ''
adj = [[] for _ in range(maxn)]
pre = [0] * maxn

def to_index(c):
    return ord(c) - ord('A')  

def to_char(index):
    return chr(index + ord('A'))  

def nhap():
    global n, m, s
    n, m, s = input().split()
    n, m = int(n), int(m)
    for _ in range(m):
        x, y, w = input().split()
        w = int(w)
        adj[to_index(x)].append((to_index(y), w))
        adj[to_index(y)].append((to_index(x), w))

def dijkstra(s):
    d = [INF] * n
    d[to_index(s)] = 0
    pre[to_index(s)] = to_index(s)

    Q = []
    heapq.heappush(Q, (0, to_index(s)))

    while Q:
        kc, u = heapq.heappop(Q)
        if kc > d[u]:
            continue
        
        for v, w in adj[u]:
            if d[v] > d[u] + w:
                d[v] = d[u] + w
                heapq.heappush(Q, (d[v], v))
                pre[v] = u

    for i in range(n):
        if to_char(i) != s:
            t = to_char(i)
            print(f"{s} -> {t}: {d[i]}")
            current = i
            path = []
            while True:
                path.append(current)
                if current == to_index(s):
                    break
                current = pre[current]
            path.reverse()
            print(" ".join(to_char(x) for x in path))

if __name__ == "__main__":
    nhap()
    dijkstra(s)

File: Bài toán mô phỏng.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

const int maxn = 26;
int n, m;
char s, t;
vector<pair<int, int>> adj[maxn];

int toIndex(char c) {
    return c - 'A';
}

char toChar(int index) {
    return index + 'A';  
}

void nhap() {
    cin >> n >> m >> s;
    for (int i = 0; i < m; i++) {
        char x, y;
        int w;
        cin >> x >> y >> w;
        adj[toIndex(x)].push_back({toIndex(y), w});
        adj[toIndex(y)].push_back({toIndex(x), w});
    }
}

const int INF = 1e9;
int pre[maxn];

void dijkstra(char s) {
    vector<ll> d(n, INF);
    d[toIndex(s)] = 0;
    pre[toIndex(s)] = toIndex(s);
    
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> Q;
    Q.push({0, toIndex(s)});
    
    while (!Q.empty()) {
        pair<int, int> top = Q.top(); Q.pop();
        int u = top.second;
        int kc = top.first;
        if (kc > d[u]) continue;

        for (auto it : adj[u]) {
            int v = it.first;
            int w = it.second;
            if (d[v] > d[u] + w) {
                d[v] = d[u] + w;
                Q.push({d[v], v});
                pre[v] = u;
            }
        }
    }

    vector<int> path;
    for (int i = 0; i < n; i++) {
        if (toChar(i) != s) {
            t = toChar(i);
            cout << "Tu " << s << " -> " << t << ": ";
            int current = i;
            while (1) {
                path.push_back(current);
                if (current == toIndex(s)) break;
                current = pre[current];
            }
            reverse(begin(path), end(path));
            for(int j = 0; j < path.size(); j++){
				cout << toChar(path[j]);
				if(j < path.size() - 1){
					cout << " - ";
				}
			}
			cout << ", cost(" << t << ") = " << d[i] << endl;
            path.clear();
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    nhap();
    dijkstra(s);
}
